File: 04 motor/servo_driver/pca9685_multi.py
#!/usr/bin/env python
#-*- coding: utf-8 -*-
#If this code works, it was written by Seunghyun Lee(www.bluebaynetworks.co.kr).
#If not, I don't know who wrote it

import smbus
import time
import math

# PCA9685 제어를 위한 레지스터 주소(데이터 시트를 참조한다.) 
PCA9685_MODE1 = 0x0
PCA9685_PRESCALE = 0xFE

LED0_ON_L = 0x6
LED0_ON_H = 0x7
LED0_OFF_L = 0x8
LED0_OFF_H = 0x9

# PCA9685 레지스터에서 1 바이트를 읽는다
def read_byte(adr):
	return bus.read_byte_data(address, adr)

# PCA9685 레지스터에 1 바이트를 기록한다
def write_byte_2c(adr, val):
	return bus.write_byte_data(address, adr, val)

# PCA9685 레지스터에 2 바이트를 기록한다
def write_word_2c(adr, val):
	bus.write_byte_data(address, adr, val)
	bus.write_byte_data(address, adr + 1, (val >> 8))

"""
아래 함수에서 진동수(frequency)에 0.9를 곱해서 사용했다.
이 문제는 PCA9685의 버그에서 비롯된다.
아래 페이지를 참조하기 바란다. 
추후 새로운 모듈에서는 이 문제가 해결될 수도 있다.
issue : https://github.com/adafruit/Adafruit-PWM-Servo-Driver-Library/issues/11
"""

# PWM 주기 설정
def set_PWMFreq(freq):
	freq *= 0.9
	prescaleval = 25000000.0
	prescaleval /= 4095
	prescaleval /= freq
	prescaleval -= 1;
	print "Estimated pre-scale: ", prescaleval

	prescale = math.floor(prescaleval + 0.5)
	print "Final pre-scale: ", prescale

	oldmode = read_byte(PCA9685_MODE1)
	newmode = (oldmode&0x7F) | 0x10
	write_byte_2c(PCA9685_MODE1, newmode)
	write_byte_2c(PCA9685_PRESCALE, int(prescale))
	write_byte_2c(PCA9685_MODE1, oldmode)
	time.sleep(0.005)
	write_byte_2c(PCA9685_MODE1, oldmode | 0xa1)
	
# PWM 설정
def set_PWM(channel, on, off):
	on = on & 0xFFFF
	off = off & 0xFFFF
	write_word_2c(LED0_ON_L+4*channel,on) 
	write_word_2c(LED0_ON_L+4*channel + 2,off) 

# PWM 듀티 설정
def set_PWM_Duty(channel, rate):
	on = 0
	off = rate * 4095.0 / 100.0 
	set_PWM(channel, on, int(off))

# SG90 모터를 좌회전(PWM 듀티 5% 설정)
def Left(start, end):
	for x in range(start, end):
		set_PWM_Duty(x, 5.0)
		
# SG90 모터를 중앙에 위치(PWM 듀티 7.5% 설정)
def Middle(start, end):
	for x in range(start, end):
		set_PWM_Duty(x, 7.5)

# SG90 모터를 우회전(PWM 듀티 10% 설정)
def Right(start, end):
	for x in range(start, end):
		set_PWM_Duty(x, 10.0)

#여기에서 부터 프로그램 시작
# I2C 통신을 위한 smbus 초기화. Revision 2 파이에서는 파라미터 1을 사용한다.
bus = smbus.SMBus(1)
# I2C 통신을 위한 PCA9685의 주소값. 점퍼 세팅을 이용해 값을 바꾸지 않았다면 0X40이 기본 값이다.
address = 0x40

# Now reset pca9685 
try:
	# PCA9685를 초기화하고 주기를 50Hz로 설정한다. SG90 모터가 50Hz에서 작동하기 때문이다.
	bus.write_byte_data(address, PCA9685_MODE1, 0)
	set_PWMFreq(50)	#50Hz
except IOError:
	print "Perhaps there's no i2c device, run i2cdetect -y 1 for device connection!" 
	pass
try:
	while True:
		Middle(0, 8)
		time.sleep(2)
		Left(0, 8)
		time.sleep(2)
		Right(0, 8)
		time.sleep(2)
		Middle(0, 8)
		time.sleep(2)
		Left(0, 4)
		time.sleep(2)
		Right(4, 8)
		time.sleep(2)

except KeyboardInterrupt:   
	print "Servo driver Application End" 
	set_PWM(0, 0, 0)
	exit()

File: 04 motor/servo_driver/pca9685_multi.c
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <math.h>
#include <wiringPi.h>
#include <wiringPiI2C.h>

#define PCA9685_MODE1 0x0
#define PCA9685_PRESCALE  0xFE

#define LED0_ON_L  0x6
#define LED0_ON_H  0x7
#define LED0_OFF_L  0x8
#define LED0_OFF_H  0x9

int dID = 0x40;	// This is the default address value of pca9685 
int fd = 0;	// i2c device handle
int Hz = 50;

void my_ctrl_c_handler(int sig){ // can be called asynchronously
	int x;

	printf ("Servo driver Application End\n"); 
	for( x  = 0; x < 8; x++)
		set_PWM(x, 0, 0);
	close(fd);
	exit(0);
}

char read_byte_2c(int addr)
{
	return (char)wiringPiI2CReadReg8(fd, addr) ;
}

void write_byte_2c(int addr, char val)
{
	wiringPiI2CWriteReg8(fd, addr, val) ;
	return;
}

void write_word_2c(int addr, short val)
{
	wiringPiI2CWriteReg8(fd, addr, val) ;
	wiringPiI2CWriteReg8(fd, addr + 1, val >> 8) ;
	return;
}

/*
Why frequency *0.9 ?
issue : https://github.com/adafruit/Adafruit-PWM-Servo-Driver-Library/issues/11
*/
void set_PWMFreq(float freq)
{
	float prescaleval, prescale;
	char oldmode, newmode, scale;

	freq *= 0.9;
	prescaleval = 25000000.0;
	prescaleval /= 4095;
	prescaleval /= freq;
	prescaleval -= 1;
	printf("Estimated pre-scale:%f\n", prescaleval);

	prescale = floor(prescaleval + 0.5);
	printf ("Final pre-scale:%f\n ", prescale);
	scale = (char)(int)prescale;
	oldmode = read_byte_2c(PCA9685_MODE1);
	newmode = (oldmode & 0x7F) | 0x10 ;
	write_byte_2c(PCA9685_MODE1, newmode);
	write_byte_2c(PCA9685_PRESCALE, scale);
	write_byte_2c(PCA9685_MODE1, oldmode);
	delay(5);
	write_byte_2c(PCA9685_MODE1, oldmode | 0xa1);
}

void set_PWM(int channel, short on, short off)
{
	on = on & 0xFFFF;
	off = off & 0xFFFF;
	write_word_2c(LED0_ON_L+4*channel,on) ;
	write_word_2c(LED0_ON_L+4*channel + 2,off) ;

}
// PWM Length 설정
void set_PWM_Length(int channel, double rate)
{
  float pulse, off;
  int on;  
  pulse = 1000.0 / Hz; //perhaps 20ms
  off = rate * 4095 / pulse;
  on = 0;
  set_PWM(channel, on, (int)off);
}

// angle 설정 (SG90의 경우 대략... 0도=0.6ms ... 180도=2.5ms)
void set_angle(int channel, double rate)
{
  double val;	
  val = 0.6 + rate * 1.9 / 180.0;
  set_PWM_Length(channel, val);
}

void Left(int start, int end)
{
	int x;
	for( x  = start; x < end; x++)
		set_angle(x, 0.0);
}		
void Middle(int start, int end)
{
	int x;
	for( x  = start; x < end; x++)
		set_angle(x, 90.0);
}
void Right(int start, int end)
{
	int x;
	for( x  = start; x < end; x++)
		set_angle(x, 180.0);
}
int main()
{
	char buf[128];
	
	signal(SIGINT, my_ctrl_c_handler);	//Ctrl + C Handler
	if((fd=wiringPiI2CSetup(dID))<0){
		printf("error opening i2c channel\n\r");
		return 0;
	}

	write_byte_2c(PCA9685_MODE1, 0);
	set_PWMFreq(Hz); //50Hz

	while(1){
		Middle(0, 8);
		sleep(2);
		Left(0, 8);
		sleep(2);
		Right(0, 8);
		sleep(2);
		Middle(0, 8);
		sleep(2);
		Left(0, 4);
		sleep(2);
		Right(4, 8);
		sleep(2);
	}
	close(fd);
	printf ("Servo driver Application End\n"); 
	set_PWM(0, 0, 0);
	return 0;
}


